_source/_posts/2020-12-08-beginners-guide-to-jwt.adoc

```diff
@@ -0,0 +1,175 @@
+
+Next, each of the first two sections are base64-url decoded:
+
+[source,json]
+----
+{"alg":"HS256"} # header
+.
+{"name":"Joe Coder"} # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: On macOS  (with "coreutils" installed, you can base64-url decode the strings on the command line with: `echo eyJhbGciOiJIUzI1NiJ9 | gbasenc -d --base64url`
+
+The last section in the JWT, the signature, is also base64-url encoded, but it's just binary data; if you try to decode it, you will end up with non-displayable characters:
+
+```txt
+��i�i�￹�￹KN�f￹�￹֪�O�_￹R��￹\�+￹
+```
+
+=== JWT Claims
+
+Once you start using JWTs you start hearing the word "claim" everywhere. A JWT claim is a key/value pair in a JSON object.  In the example above, `"name": "Joe Coder"`, the claim key is 'name' and the value is 'Joe Coder`. The value of a claim can be any JSON object.
+
+There are three types of claims: "registered," "public," and "private." You can find the list of registered and public claims in the https://www.iana.org/assignments/jwt/jwt.xhtml#claims[official IANA Registry].  You can also add any other custom claim to a JWT; these are known as "private claims."
+
+WARNING: When using private claims, watch out for name collisions with the official claims.
+
+The use of registered claims is optional, but when they are present, they MUST be validated.  For example, a JWT may contain date-time fields that describe when the token is valid.
+
+- Issued At (`iat`) - The time the JWT was created
+- Expiration Time (`exp`) - The time at which the JWT is no longer valid
+- Not Before (`nfb`) - The earliest time the JWT would be valid
+
+NOTE: Timestamps are "seconds since the epoch" integer format. JWT libraries can add up to a few minutes of leeway to these values to account for clock skew between systems.
+
+=== JWT Header
+
+The header of a JWT contains information about how the token was created.  In my example, the "algorithm" (`alg`) claim is set to `HS256`, which specifies the hashing algorithm HMAC SHA-256 is used to generate or validate the signature.
+
+[source,json]
+----
+{
+  "alg": "HS256"
+}
+----
+
+=== JWT Signature
+
+The JWT specifications list a few different signing algorithms; each of these algorithms work slightly different. For simplicity's sake, there are two types of algorithms:
+- HMAC based shared secret, these all start with the prefix `HS`, which stands for HMAC SHA)
+- Public key pair (either RSA or ECDSA keys)
+
+WARNING: Caution is needed when using a shared secret, as anyone with the secret can create (or forge) new JWTs. If you need to validate a JWT from an untrusted client (web-page, mobile app, etc.), use a public key pair instead.
+
+The JWT in this example (actually a JWS, remember the `S` stands for "signature") uses the `HS256` algorithm.  To validate the signature, the hash of the first two parts of the token is rehashed, and then output it as a base64-url string.
```

Comment:
"To validate the signature, start by base64-url decoding the signature. Then HMAC of the first two parts of the token and finally, compare the output of your HMAC operation with the results of the base64-url decoded signature"